src/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -78,22 +86,51 @@ instance Eq (ReadCtx fhandle) where
 instance Ord (ReadCtx fhandle) where
   compare = compare `on` (\r -> (readCtxHeadKey r, readCtxNumber r))
 
+-- TODO: This is slightly inelegant. This module could work generally for
+-- anything that can produce elements, but currently is very specific to having
+-- write buffer and run readers. Also, for run merging, no write buffer is
+-- involved, but we still need to branch on this sum type.
+-- A more general version is possible, but despite SPECIALISE-ing everything
+-- showed ~100 bytes of extra allocations per entry that is read (which might be
+-- avoidable with some tinkering).
+data Reader m fhandle =
+    ReadRun    !(RunReader m fhandle)
+    -- The list allows to incrementally read from the write buffer without
+    -- having to find the next entry in the Map again (requiring key
+    -- comparisons) or having to copy out all entries.
+    -- TODO: more efficient representation? benchmark!
+  | ReadBuffer !(MutVar (PrimState m) [KOp m fhandle])
```

Comment:
This causes [NoThunks test failures], but a thunk probably makes sense here. Should I somehow change the `NoThunks` instance to allow a thunk in this particular place? @jorisdral

Reply:
I'll have a look

Reply:
For now, let's just do this:

```diff
diff --git a/src-extras/Database/LSMTree/Extras/NoThunks.hs b/src-extras/Database/LSMTree/Extras/NoThunks.hs
index cee8a61c..597071c7 100644
--- a/src-extras/Database/LSMTree/Extras/NoThunks.hs
+++ b/src-extras/Database/LSMTree/Extras/NoThunks.hs
@@ -525,7 +525,9 @@ instance NoThunks a => NoThunks (MutVar s a) where
   showTypeOf (_ :: Proxy (MutVar s a)) = "MutVar"
   wNoThunks ctx var = do
       x <- unsafeSTToIO $ readMutVar var
-      noThunks ctx x
+      -- TODO: disallow thunk once we find out why atomicModifyMutVar' is
+      -- creating thunks
+      noThunks ctx (AllowThunk x)

 -- TODO: https://github.com/input-output-hk/nothunks/issues/56
 deriving via OnlyCheckWhnf (PrimVar s a)
```

I might have to play around with `MutVar`s to find out why `atomicModifyMutVar'` is creating thunks...

Reply:
I don't think it is about the `MutVar` here, but the list inside of it. It is intentionally a thunk. We don't want to copy the whole write buffer into a list, but incrementally evaluate the list as we read entries.

I will try tweaking the instance `NoThunks (Reader m (Handle h))`.

Reply:
Like this?


```haskell
instance (Typeable (PrimState m), Typeable h)
      => NoThunks (Reader m (Handle h)) where
  showTypeOf (_ :: Proxy (Reader m (Handle h))) = "Reader"
  wNoThunks ctx = \case
    ReadRun r      -> noThunks ctx r
    ReadBuffer var -> return Nothing  -- intentionally lazy
```

Reply:
It doesn't check that the MutVar itself is not a thunk, as that seemed slightly trickier to do without also checking the content of the MutVar. But maybe good enough?

Reply:
```diff
- ReadBuffer var -> return Nothing  -- intentionally lazy
+ ReadBuffer var -> noThunks ctx (OnlyCheckWhnf var)